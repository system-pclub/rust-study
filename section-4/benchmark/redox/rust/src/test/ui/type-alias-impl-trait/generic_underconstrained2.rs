#![feature(type_alias_impl_trait)]

fn main() {}

type Underconstrained<T: std::fmt::Debug> = impl 'static;
//~^ ERROR `U` doesn't implement `std::fmt::Debug`
//~^^ ERROR: at least one trait must be specified

// not a defining use, because it doesn't define *all* possible generics
fn underconstrained<U>(_: U) -> Underconstrained<U> {
    5u32
}

type Underconstrained2<T: std::fmt::Debug> = impl 'static;
//~^ ERROR `V` doesn't implement `std::fmt::Debug`
//~^^ ERROR: at least one trait must be specified

// not a defining use, because it doesn't define *all* possible generics
fn underconstrained2<U, V>(_: U, _: V) -> Underconstrained2<V> {
    5u32
}
