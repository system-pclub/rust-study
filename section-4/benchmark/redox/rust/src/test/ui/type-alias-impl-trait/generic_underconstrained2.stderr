error: at least one trait must be specified
  --> $DIR/generic_underconstrained2.rs:5:45
   |
LL | type Underconstrained<T: std::fmt::Debug> = impl 'static;
   |                                             ^^^^^^^^^^^^

error: at least one trait must be specified
  --> $DIR/generic_underconstrained2.rs:14:46
   |
LL | type Underconstrained2<T: std::fmt::Debug> = impl 'static;
   |                                              ^^^^^^^^^^^^

error[E0277]: `U` doesn't implement `std::fmt::Debug`
  --> $DIR/generic_underconstrained2.rs:5:1
   |
LL | type Underconstrained<T: std::fmt::Debug> = impl 'static;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `U` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
...
LL | fn underconstrained<U>(_: U) -> Underconstrained<U> {
   |                     - help: consider restricting this bound: `U: std::fmt::Debug`
   |
   = help: the trait `std::fmt::Debug` is not implemented for `U`
   = note: the return type of a function must have a statically known size

error[E0277]: `V` doesn't implement `std::fmt::Debug`
  --> $DIR/generic_underconstrained2.rs:14:1
   |
LL | type Underconstrained2<T: std::fmt::Debug> = impl 'static;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `V` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
...
LL | fn underconstrained2<U, V>(_: U, _: V) -> Underconstrained2<V> {
   |                         - help: consider restricting this bound: `V: std::fmt::Debug`
   |
   = help: the trait `std::fmt::Debug` is not implemented for `V`
   = note: the return type of a function must have a statically known size

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0277`.
